src/lint.cpp

```diff
@@ -423,16 +422,23 @@ void linter::propagate_variable_declarations_to_parent_scope() {
 }
 
 void linter::report_error_if_assignment_is_illegal(
-    const linter::declared_variable *var, const identifier &assignment) const {
+    const linter::declared_variable *var, const identifier &assignment, bool is_assigned_before_declaration) const {
   switch (var->kind) {
     case variable_kind::_const:
     case variable_kind::_import:
       if (var->is_global_variable) {
         this->error_reporter_->report(
             error_assignment_to_const_global_variable{assignment});
       } else {
-        this->error_reporter_->report(error_assignment_to_const_variable{
-            var->declaration(), assignment, var->kind});
+        if (!is_assigned_before_declaration) {
```

Comment:
Nit: I dislike double negatives; line 438 executes if `! !is_assigned_before_declaration`. Invert this `if` to avoid double negatives.